### PR TITLE
PKCS11: Test Cases for Object APIs - Finding Objects

### PR DESCRIPTION
**Pull request by ruchi393**
Add 1011 for -
C_FindObjectsInit()
C_FindObjects()
C_FindObjectsFinal()

Signed-off-by: Ruchika Gupta <ruchika.gupta@linaro.org>

### CODE REVIEW

**Comment by ruchi393** [comment]
This is dependent on following PR's
https://github.com/OP-TEE/optee_os/pull/4286
https://github.com/OP-TEE/optee_os/pull/4288
https://github.com/OP-TEE/optee_client/pull/244

File: host/xtest/pkcs11_1000.c [hunk 1]
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+	CK_ATTRIBUTE find_template[] = {
+		{ CKA_LABEL, (CK_UTF8CHAR_PTR)label, strlen(label) },
+	};
+	CK_ATTRIBUTE find_token_template[] = {
+		{ CKA_LABEL, (CK_UTF8CHAR_PTR)label, strlen(label) },
+		{ CKA_TOKEN, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+	CK_ATTRIBUTE find_session_template[] = {
+		{ CKA_LABEL, (CK_UTF8CHAR_PTR)label, strlen(label) },
+		{ CKA_TOKEN, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
```
Comment (vesajaaskelainen) [inline host/xtest/pkcs11_1000.c:2236]: To exercise the serializer also a bit it is a good idea to have CK_ULONG attributes (as they have special handling) and CK_BBOOL attributes (they are only 1 byte) in between there.

File: host/xtest/pkcs11_1000.c [hunk 2]
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+
+	rv = C_FindObjectsFinal(session);
+	ADBG_EXPECT_CK_OK(c, rv);
+
+	rv = C_FindObjectsInit(session, find_template,
+			       ARRAY_SIZE(find_template));
+	ADBG_EXPECT_CK_OK(c, rv);
+
+	/*
+	 * Intentionlly do not finalize the active object search. It should
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:2513]: s/Intentionlly/Intentionally/

File: host/xtest/pkcs11_1000.c (hunk 1 above)
Comment (etienne-lms) [reply host/xtest/pkcs11_1000.c:2236]: Maybe add specific extra tests in a later change?

**Review by etienne-lms — COMMENTED** [review]
~~`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`~~
(edited: removed this content changes)

File: host/xtest/pkcs11_1000.c (hunk 1 above)
Comment (vesajaaskelainen) [reply host/xtest/pkcs11_1000.c:2236]: I am OK with that. When we have set and copy support it is much easier to make all kind of tests for attributes.

Comment (ruchi393) [reply host/xtest/pkcs11_1000.c:2236]: Thanks @etienne-lms , @vesajaaskelainen . I will take a note and add a case in it with set and copy support.

File: host/xtest/pkcs11_1000.c [hunk 3]
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+	/*
+	 * Token Private Obj CKA_TOKEN = CK_TRUE, CKA_PRIVATE = CK_TRUE
+	 * Expected error as User not logged in
+	 */
+	rv = create_data_object(session, &obj_hdl[object_id++], CK_TRUE,
+				CK_TRUE, label);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_USER_NOT_LOGGED_IN, rv))
+		goto out;
+
+	object_id--;
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:2287]: should remove this line and `++` incremention from line 2282.

File: host/xtest/pkcs11_1000.c [hunk 4]
```diff
@@ -2312,6 +2314,17 @@ static void xtest_pkcs11_test_1011(ADBG_Case_t *c)
 	if (!ADBG_EXPECT_CK_OK(c, rv))
 		goto out;
 
+	/*
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:2317]: You forgot to move `Do_ADBG_BeginSubCase()`/`Do_ADBG_EndSubCase()`.

Comment (ruchi393) [reply host/xtest/pkcs11_1000.c:2317]: I added this as a check to ensure all handles are valid. Its not a new Subcase.

Comment (etienne-lms) [reply host/xtest/pkcs11_1000.c:2317]: Ok, then, should remove sub test 9 that is now empty.

Comment (ruchi393) [reply host/xtest/pkcs11_1000.c:2317]: Sorry for that. Already added a commit for same.

File: host/xtest/pkcs11_1000.c [hunk 5]
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+}
+
+static void xtest_pkcs11_test_1011(ADBG_Case_t *c)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_SLOT_ID slot = 0;
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+	CK_OBJECT_HANDLE obj_hdl[10];
+	CK_OBJECT_HANDLE obj_found[10];
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply host/xtest/pkcs11_1000.c:2221]: not addressed

File: host/xtest/pkcs11_1000.c [hunk 6]
```diff
@@ -2114,3 +2114,441 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+			       ARRAY_SIZE(find_template),
+			       obj_found, ARRAY_SIZE(obj_found), 4);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	/*
+	 * Check if object handles returned when creating objects with this
+	 * session are still valid
+	 */
+	for (uint32_t i = 0; i < object_id; i++) {
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:2321]: Prefer `uint32_t i = 0;` definition at function entry.
See also the other `for()` loop.

File: host/xtest/pkcs11_1000.c [hunk 7]
```diff
@@ -2114,3 +2114,441 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+
+	rv = C_Logout(session);
+	ADBG_EXPECT_CK_OK(c, rv);
+
+	logged_in = false;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	/*
+	 * Sub test 8: Close session and open new session , find objects
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:2468]: space char before coma.

File: host/xtest/pkcs11_1000.c (hunk 6 above)
Comment (ruchi393) [reply host/xtest/pkcs11_1000.c]: Done

File: host/xtest/pkcs11_1000.c (hunk 7 above)
Comment (ruchi393) [reply host/xtest/pkcs11_1000.c]: Fixed

File: host/xtest/pkcs11_1000.c (hunk 2 above)
Comment (ruchi393) [reply host/xtest/pkcs11_1000.c]: Done

File: host/xtest/pkcs11_1000.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ruchi393) [reply host/xtest/pkcs11_1000.c]: Sorry, missed this earlier. Have fixed it now.

File: host/xtest/pkcs11_1000.c (hunk 3 above)
Comment (ruchi393) [reply host/xtest/pkcs11_1000.c]: Done

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

Thanks @etienne-lms . I will squash commits and apply tags once the optee_os and optee_client patches are accepted.

File: host/xtest/pkcs11_1000.c [hunk 8]
```diff
@@ -2114,3 +2114,442 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
...
+
+	/* Session Private Obj CKA_TOKEN = CK_FALSE, CKA_PRIVATE = CK_TRUE */
+	rv = create_data_object(session, &obj_hdl[object_id++], CK_FALSE,
+				CK_TRUE, label);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = test_find_objects(c, session, find_template,
+			       ARRAY_SIZE(find_template),
+			       obj_found, ARRAY_SIZE(obj_found), 4);
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:2314]: These line could be change to test that providing no attribute return all visible objects:
```patch
-	rv = test_find_objects(c, session, find_template,
-			       ARRAY_SIZE(find_template),
+	rv = test_find_objects(c, session, NULL, 0,
 			       obj_found, ARRAY_SIZE(obj_found), 4);
```

Comment (ruchi393) [reply host/xtest/pkcs11_1000.c:2314]: @etienne-lms , saw your comment late. I just added another sub-case for this.

**Comment by etienne-lms** [comment]
My `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` still stands.

**Comment by ruchi393** [comment]
Squashed, rebased and tags applied. 

This PR has a dependency on 
https://github.com/OP-TEE/optee_os/pull/4288
https://github.com/OP-TEE/optee_client/pull/244

**Comment by jforissier** [comment]
> Squashed, rebased and tags applied.
> 
> This PR has a dependency on
> [OP-TEE/optee_os#4288](https://github.com/OP-TEE/optee_os/pull/4288)
> [OP-TEE/optee_client#244](https://github.com/OP-TEE/optee_client/pull/244)

All, do you think it's OK for me to merge those PRs once they are properly reviewed, even though we're post 3.12.0-rc1? My opinion is: yes it is, since PKCS#11 is still experimental and incomplete so the risk of regression is quite minimal.

**Comment by etienne-lms** [comment]
Hi Jérôme, I think it is safe too, for the reasons you gave.

**Comment by etienne-lms** [comment]
... but of course let's wait OP-TEE/optee_os#4288 & OP-TEE/optee_client#244 have landed before merging this  optee_test P-R :)

**Comment by vesajaaskelainen** [comment]
Did quick look and did not spot anything alarming. If you want to go ahead with merge feel free and no need to wait me for doing better review for test cases here.

**Comment by ruchi393** [comment]
Thanks @vesajaaskelainen .

@jforissier , we can probably merge the  OP-TEE/optee_os#4288 & OP-TEE/optee_client#244 now.  Both those PR's have R-by tags applied now.  That would also fix the  IBART errors in this PR and we can consider merging it.

**Comment by jforissier** [comment]
@ruchi393 yes I agree, I have seen the updates and I'm just waiting for CI to finish before merging. Thanks.

**Pull request merged**